src/Storages/StorageDistributed.cpp

```diff
@@ -437,11 +437,31 @@ StoragePtr StorageDistributed::createWithOwnCluster(
     return res;
 }
 
-QueryProcessingStage::Enum StorageDistributed::getQueryProcessingStage(const Context &context, QueryProcessingStage::Enum to_stage, const ASTPtr & query_ptr) const
+QueryProcessingStage::Enum StorageDistributed::getQueryProcessingStage(const Context & context, QueryProcessingStage::Enum to_stage, const SelectQueryInfo & query_info) const
```

Comment:
Why we cannot drop `const` in SelectQueryInfo to avoid confusion?

Reply:
Did it my analogy with `input_order_info` which had mutable qualifier and for simplicity (to avoid possible conflicts and to make review easier), but agree this should be cleaned up

FWIW, done in a separate commit